src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java

```diff
@@ -0,0 +1,209 @@
+import uk.gov.hmcts.reform.bulkscanprocessor.model.common.Event;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.document.output.Pdf;
+import uk.gov.hmcts.reform.bulkscanprocessor.tasks.processor.BlobManager;
+import uk.gov.hmcts.reform.bulkscanprocessor.tasks.processor.DocumentProcessor;
+import uk.gov.hmcts.reform.bulkscanprocessor.tasks.processor.EnvelopeProcessor;
+import uk.gov.hmcts.reform.bulkscanprocessor.tasks.processor.ZipFileProcessingResult;
+import uk.gov.hmcts.reform.bulkscanprocessor.tasks.processor.ZipFileProcessor;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.Optional;
+import java.util.zip.ZipInputStream;
+
+import static java.time.Instant.now;
+import static java.time.temporal.ChronoUnit.MINUTES;
+import static java.util.stream.Collectors.groupingBy;
+import static org.slf4j.LoggerFactory.getLogger;
+import static uk.gov.hmcts.reform.bulkscanprocessor.entity.Status.CREATED;
+import static uk.gov.hmcts.reform.bulkscanprocessor.model.common.Event.DOC_FAILURE;
+import static uk.gov.hmcts.reform.bulkscanprocessor.model.common.Event.DOC_SIGNATURE_FAILURE;
+import static uk.gov.hmcts.reform.bulkscanprocessor.model.common.Event.DOC_UPLOADED;
+import static uk.gov.hmcts.reform.bulkscanprocessor.model.common.Event.DOC_UPLOAD_FAILURE;
+
+/**
+ * Service responsible to upload envelopes ended in state after main processor task.
+ * <p></p>
+ * {@link uk.gov.hmcts.reform.bulkscanprocessor.tasks.BlobProcessorTask}
+ * <p></p>
+ * All the validations have been accomplished and {@link Envelope} saved in DB.
+ * Therefore no need to go through the same set of validation steps/rules and just re-extract zip file and upload.
+ * In case some {@link StorageException} or {@link IOException} experienced during zip extraction
+ * the envelope state will not be changed and left for a retry on the next run.
+ * {@link Event#DOC_UPLOAD_FAILURE} can be treated as stuck envelope and be re-uploaded by same service here.
+ * There is an upload retry counter/limit which may be used as well
+ * Review ^ last sentence and, perhaps, incorporate such possibility here.
+ */
+@Service
+public class UploadEnvelopeDocumentsService {
+
+    private static final Logger log = getLogger(UploadEnvelopeDocumentsService.class);
+
+    private final long coolDownPeriod;
+    private final EnvelopeRepository envelopeRepository;
+    private final BlobManager blobManager;
+    private final ZipFileProcessor zipFileProcessor;
+    private final ProcessEventRepository eventRepository;
+    private final DocumentProcessor documentProcessor;
+    private final EnvelopeProcessor envelopeProcessor;
+
+    public UploadEnvelopeDocumentsService(
+        // only process envelopes after cool down period
+        // can be removed once uploading feature is removed from main job
+        @Value("${scheduling.task.upload-documents.cool-down-minutes}") long coolDownPeriod,
```

Comment:
👍

Reply:
How about `minimumEnvelopeAge` or sth? It wasn't clear to me what `coolDownPeriod` meant.

Reply:
Btw very well thought with this parameter